flax/cursor.py

```diff
@@ -0,0 +1,304 @@
+import enum
+# from sqlite3 import Cursor
+from typing import Any, Callable, Generic, List, Mapping, Optional, Protocol, Sequence, Tuple, TypeVar, Union, runtime_checkable
+import jax
+from flax.core import FrozenDict
+import dataclasses
+
+
+A = TypeVar('A')
+Key = Any
+
+
+@runtime_checkable
+class Indexable(Protocol):
+
+  def __getitem__(self, key) -> Any:
+    ...
+
+
+class AccessType(enum.Enum):
+  GETATTR = enum.auto()
+  GETITEM = enum.auto()
+
+
+@dataclasses.dataclass
+class ParentKey(Generic[A]):
+  parent: 'Cursor[A]'
+  key: Any
+
+
+class Cursor(Generic[A]):
+  obj: A
+  parent_key: Optional[ParentKey[A]]
+  changes: dict[Any, Union[Any, 'Cursor[A]']]
+
+  def __init__(self, obj: A, parent_key: Optional[ParentKey[A]]):
+    if isinstance(obj, (FrozenDict, dict, list, tuple)):
+      obj_copy = jax.tree_util.tree_map(lambda x: x, obj)
+    elif dataclasses.is_dataclass(obj):
+      obj_copy = dataclasses.replace(obj)
+    else:
+      obj_copy = obj
+
+    # NOTE: we use `vars` here to avoid calling `__setattr__`
+    # vars(self) = self.__dict__
+    vars(self)['obj'] = obj_copy
+    vars(self)['parent_key'] = parent_key
+    vars(self)['changes'] = {}
+
+  @property
+  def root(self) -> 'Cursor[A]':
+    if self.parent_key is None:
+      return self
+    else:
+      return self.parent_key.parent.root
+
+  def __getitem__(self, key) -> 'Cursor[A]':
+    if key in self.changes:
+      return self.changes[key]
+
+    if not isinstance(self.obj, Indexable):
+      raise TypeError(f'Cannot index into {self.obj}')
+
+    if isinstance(self.obj, Mapping) and key not in self.obj:
+      raise KeyError(f'Key {key} not found in {self.obj}')
+
+    child = Cursor(self.obj[key], ParentKey(self, key))
+    self.changes[key] = child
+    return child
+
+  def __getattr__(self, name) -> 'Cursor[A]':
+    if name in self.changes:
+      return self.changes[name]
+
+    if not hasattr(self.obj, name):
+      raise AttributeError(f'Attribute {name} not found in {self.obj}')
+
+    child = Cursor(getattr(self.obj, name), ParentKey(self, name))
+    self.changes[name] = child
+    return child
+
+  def __setitem__(self, key, value):
+    self.changes[key] = Cursor(value, ParentKey(self, key))
+
+  def __setattr__(self, name, value):
+    self.changes[name] = Cursor(value, ParentKey(self, name))
+
+  def apply_filter(
+      self,
+      filter_fn: Callable[[Sequence[str], Any], Tuple[bool, Any]],
+      _path: List[str] = [],
```

Comment:
Make `_path` a `tuple[str, ...] = ()`. Its usually not a good idea to have lists as a default value because its not immutable.